### PR TITLE
Support case expression

### PR DESCRIPTION
**Pull request by jievince**
http://jira.nebula-graph.io:8081/projects/NG/issues/NG-281?filter=allissues
Based on:
https://github.com/vesoft-inc/nebula-graph/pull/353
https://github.com/vesoft-inc/nebula-storage/pull/183

### CODE REVIEW

File: src/common/expression/CaseExpression.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    explicit CaseExpression(Kind kind, CaseList* cases) : Expression(kind) {
+        cases_ = std::move(*cases).items();
+        delete cases;
+    }
+
+    bool operator==(const Expression& rhs) const override;
+
+    const Value& eval(ExpressionContext& ctx) override;
+
+    using Item = std::pair<std::unique_ptr<Expression>, std::unique_ptr<Expression>>;
```
Comment (CPWstatic) [inline src/common/expression/CaseExpression.h:47]: Make it public in `CaseList`, then you can use it as `CaseList::Item`

File: src/common/expression/Expression.cpp [hunk 2]
```diff
@@ -593,6 +609,15 @@ std::ostream& operator<<(std::ostream& os, Expression::Kind kind) {
         case Expression::Kind::kLabel:
             os << "Label";
             break;
+        case Expression::Kind::kSimpleCase:
+            os << "SimpleCase";
+            break;
+        case Expression::Kind::kGenericCase:
+            os << "GenericCase";
+            break;
+        case Expression::Kind::kConditional:
```
Comment (CPWstatic) [inline src/common/expression/Expression.cpp:618]: kConditionalCase

File: src/common/expression/CaseExpression.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            }
+            break;
+        }
+        case Kind::kGenericCase:
+        case Kind::kConditional: {
+            for (const auto& whenThen : cases_) {
+                auto when = whenThen.first->eval(ctx);
+                auto then = whenThen.second->eval(ctx);
+                if (!when.isBool()) {
+                    LOG(FATAL) << "not bool";
```
Comment (CPWstatic) [inline src/common/expression/CaseExpression.cpp:77]: Return `null`.

Comment (CPWstatic) [reply src/common/expression/CaseExpression.cpp:77]:
```
if (!when.isBool()) {
    return Value::kNullBadType;
} else {
    result_ = whenThen.second->eval(ctx);
    return result_;
}
```

File: src/common/expression/CaseExpression.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    void add(Expression* when, Expression* then) {
+        items_.emplace_back(when, then);
+    }
+
+    auto items() && {
+        return std::move(items_);
+    }
+
+private:
+    using Item = std::pair<std::unique_ptr<Expression>, std::unique_ptr<Expression>>;
```
Comment (CPWstatic) [inline src/common/expression/CaseExpression.h:30]: Make it a struct.

File: src/common/expression/CaseExpression.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                }
+                if (when.getBool()) {
+                    result_ = then;
+                    return result_;
+                }
+            }
+            if (default_) {
+                result_ = default_->eval(ctx);
+            } else {
+                result_ = Value::kEmpty;
```
Comment (CPWstatic) [inline src/common/expression/CaseExpression.cpp:87]:
```
 if there is no ELSE case and no match is found, null will be returned.
```

According to the doc, we should return `null` here.

File: src/common/expression/CaseExpression.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        case Kind::kGenericCase:
+        case Kind::kConditionalCase: {
+            for (const auto& whenThen : cases_) {
+                auto when = whenThen.when->eval(ctx);
+                auto then = whenThen.then->eval(ctx);
+                if (!when.isBool()) {
+                    return Value::kNullBadType;
+                }
+                if (when.getBool()) {
```
Comment (laura-ding) [inline src/common/expression/CaseExpression.cpp:79]: the if is redundant

File: src/common/expression/CaseExpression.h [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+private:
+    void writeTo(Encoder& encoder) const override;
+
+    void resetFrom(Decoder& decoder) override;
+
+    std::unique_ptr<Expression> condition_{nullptr};
+    std::unique_ptr<Expression> default_{nullptr};
+    std::vector<CaseList::Item> cases_;
```
Comment (laura-ding) [inline src/common/expression/CaseExpression.h:124]: why not directly use `CaseList`?

File: src/common/expression/Expression.cpp [hunk 8]
```diff
@@ -452,13 +433,27 @@ std::unique_ptr<Expression> Expression::decode(Expression::Decoder& decoder) {
...
+            exp = std::make_unique<CaseExpression>(Expression::Kind::kGenericCase);
+            exp->resetFrom(decoder);
+            return exp;
+        }
+        case Expression::Kind::kConditionalCase: {
+            exp = std::make_unique<CaseExpression>(Expression::Kind::kConditionalCase);
+            exp->resetFrom(decoder);
+            return exp;
+        }
+            // no default so the compiler will warning when lack
```
Comment (laura-ding) [inline src/common/expression/Expression.cpp:451]: indentation

File: src/common/expression/Expression.cpp [hunk 9]
```diff
@@ -617,8 +612,17 @@ std::ostream& operator<<(std::ostream& os, Expression::Kind kind) {
...
             os << "Label";
             break;
+        case Expression::Kind::kSimpleCase:
+            os << "SimpleCase";
+            break;
+        case Expression::Kind::kGenericCase:
+            os << "GenericCase";
+            break;
+        case Expression::Kind::kConditionalCase:
+            os << "Conditional";
```
Comment (laura-ding) [inline src/common/expression/Expression.cpp:622]: ConditionalCase

File: src/common/expression/test/ExpressionTest.cpp [hunk 10]
```diff
@@ -1327,26 +1309,26 @@ TEST_F(ExpressionTest, NotInList) {
 TEST_F(ExpressionTest, NotInSet) {
     {
         auto *elist = new ExpressionList;
-        (*elist).add(new ConstantExpression(12345))
-                .add(new ConstantExpression("Hello"))
-                .add(new ConstantExpression(true));
+        (*elist)
```
Comment (laura-ding) [inline src/common/expression/test/ExpressionTest.cpp:1312]: why change the format?

**Review by laura-ding — COMMENTED** [review]
add test about `decode` and `encode`

File: src/common/expression/CaseExpression.cpp (hunk 6 above)
Comment (jievince) [reply src/common/expression/CaseExpression.cpp:79]: Why? First `if` is to check whether when is of bool type, second `if` is to get the value of when.

File: src/common/expression/CaseExpression.h (hunk 7 above)
Comment (jievince) [reply src/common/expression/CaseExpression.h:124]: if use CaseList, it will be CaseExpression.cases().items() when referring to items. It's too long.

Comment (Shylock-Hg) [reply src/common/expression/CaseExpression.h:124]: Agree I think CaseList is better.

**Review by dutor — APPROVED** [review]
Awesome work

File: src/common/expression/CaseExpression.h (hunk 7 above)
Comment (jievince) [reply src/common/expression/CaseExpression.h]: If use CaseList, both CaseList and CaseExpression will need to implement add(), setWhen() and setThen(). It's going to look a little bloated.

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**